test/cloudnative/agentversion_set.go

```diff
@@ -0,0 +1,183 @@
+	setup.InstallDynatraceFromSource(agentVersionset, &secretConfig)
+
+	setup.AssessOperatorDeployment(agentVersionset)
+
+	agentVersionset.Assess("install dynakube", dynakube.Apply(dk))
+
+	assessVersionChecks(agentVersionset, oldVersion)
+
+	agentVersionset.Assess("update dynakube", updateDynakube(newVersion))
+	agentVersionset.Assess("dynakube phase changes to 'Running'",
+		dynakube.WaitForDynakubePhase(dynakube.NewBuilder().WithDefaultObjectMeta().Build()))
+	setup.AssessOperatorDeployment(agentVersionset)
+
+	assessVersionChecks(agentVersionset, newVersion)
+
+	return agentVersionset.Feature()
+}
+
+func getAvailableVersions(secret secrets.Secret, t *testing.T) []string {
+	dtc, err := dtclient.NewClient(secret.ApiUrl, secret.ApiToken, secret.ApiToken)
+	require.NoError(t, err)
+	versions, err := dtc.GetAgentVersions(dtclient.OsUnix, dtclient.InstallerTypeDefault, arch.FlavorDefault, arch.Arch)
+	require.NoError(t, err)
+
+	return versions
+}
+
+func assessVersionChecks(builder *features.FeatureBuilder, version version.SemanticVersion) {
+	builder.Assess("restart csi driver", restartCSIDriver)
+	builder.Assess("start sample apps and injection", sampleapps.Restart)
+	builder.Assess("check init containers", checkInitContainers)
+	builder.Assess("check env vars of init container", checkVersionInSampleApp(version))
+}
+
+// this method assigns the queried versions to the variables 'old' and 'new'.
+// it makes sure, that old gets an older version than new, in order to
+// be able to simulate the upgrade of version.
+func assignVersions(t *testing.T, versions []string, old version.SemanticVersion, new version.SemanticVersion) (version.SemanticVersion, version.SemanticVersion) {
+	first := versions[0]
+	second := versions[len(versions)-1]
+
+	firstVersion, err := version.ExtractSemanticVersion(first)
+	require.NoError(t, err)
+	secondVersion, err := version.ExtractSemanticVersion(second)
+	require.NoError(t, err)
+
+	compare := version.CompareSemanticVersions(firstVersion, secondVersion)
+	if compare > 0 {
+		old = secondVersion
+		new = firstVersion
+	} else if compare < 0 {
+		old = firstVersion
+		new = secondVersion
+	}
+	return old, new
+}
+
+func updateDynakube(semanticVersion version.SemanticVersion) features.Func {
+	return func(ctx context.Context, t *testing.T, environmentConfig *envconf.Config) context.Context {
+		resources := environmentConfig.Client().Resources()
+		var dk dynatracev1beta1.DynaKube
+
+		err := dynatracev1beta1.AddToScheme(resources.GetScheme())
+		require.NoError(t, err)
+
+		err = resources.Get(ctx, "dynakube", "dynatrace", &dk)
+		require.NoError(t, err)
+
+		dk.Status.UpdatedTimestamp = metav1.Now()
+		dk.Spec.OneAgent.CloudNativeFullStack.Version = semanticVersion.String()
+		err = resources.Update(ctx, &dk)
+		require.NoError(t, err)
+
+		return ctx
+	}
+}
+
+func restartCSIDriver(ctx context.Context, t *testing.T, environmentConfig *envconf.Config) context.Context {
```

Comment:
why restart the csi driver?

Reply:
without this the injection does not take place. i am not sure why, but when i apply a dynakube with a specific version i see that the csi pods do restart (i thought maybe this is because they already have the latest version and when i apply a dynakube with a different version it comes to problems thats why i decided to restart it simply once after applying the dynakube and then it works fine)

Reply:
if you have another, better way to solve this, i would happily change this

Reply:
that sounds like something that should not happen 😅

Reply:
alright i'll think of ways to do this better. thanks